### PR TITLE
1297: Add CMake variable to set max number of ranks/nodes for tests and examples.

### LINKED ISSUE #1297: Expand unit tests to run on more nodes and also run on just one node
**What Needs to be Done?**

- Oversubscribe the containers to test more nodes
- Add single rank units test runs (with potentially some exclusion)

### PR DESCRIPTION
**Pull request by JacobDomagala**
Fixes #1297

### CODE REVIEW

File: cmake/load_mpi_package.cmake [hunk 1]
```diff
@@ -16,9 +16,15 @@ set(MPI_RUN_COMMAND  "${MPIEXEC_EXECUTABLE}")
 set(MPI_PRE_FLAGS    "${MPIEXEC_PREFLAGS}")
 set(MPI_EPI_FLAGS    "${MPIEXEC_POSTFLAGS}")
 set(MPI_NUMPROC_FLAG "${MPIEXEC_NUMPROC_FLAG}")
-set(MPI_MAX_NUMPROC  "${MPIEXEC_MAX_NUMPROCS}")
+set(MPI_MAX_NUMPROC  "${vt_tests_num_nodes}")
 
-message(STATUS "MPI detected max nproc: ${MPIEXEC_MAX_NUMPROCS}")
+# If requested numproc is greater than the detected one
+# we have to set --oversubscribe flag
+if(${MPI_MAX_NUMPROC} GREATER ${MPIEXEC_MAX_NUMPROCS})
```
Comment (lifflander) [inline cmake/load_mpi_package.cmake:23]: I don't think this is needed? Maybe on some MPI implementations... but I've never had to pass that

Comment (JacobDomagala) [reply cmake/load_mpi_package.cmake:23]: For MPICH it's not needed ( and also it's not supported from what i can see ) but for OpenMP workflow I was getting the following error: 
image

File: tests/CMakeLists.txt [hunk 2]
```diff
@@ -103,7 +103,7 @@ function(add_unit_test unit_test_name unit_test_files uses_mpi)
 
       set_tests_properties(
         ${${CUR_TEST_LIST}}
-        PROPERTIES TIMEOUT 60
+        PROPERTIES TIMEOUT 360
```
Comment (JacobDomagala) [inline tests/CMakeLists.txt:106]: Needed for following tests that would timeout:
image

**Comment by JacobDomagala** [comment]
Some of the tests are failing on `vt:*TestLocationRoute/*test_entity_cache_migrated_entity_proc_4` with the assertion:
image

File: tests/CMakeLists.txt (hunk 2 above)
Comment (JacobDomagala) [reply tests/CMakeLists.txt:106]: With https://github.com/DARMA-tasking/vt/pull/1329 timeout can be changed back to 60 sec

**Comment by JacobDomagala** [comment]
image

Weird, on other workflows this one is finishing in less than a second. Will see locally, if enabling sanitizer will reproduce it

**Comment by JacobDomagala** [comment]
I've set all workflows to run on 4 nodes now, but I'm not sure if we want to keep it that way, since the test execution increased 2 times (at least)

**Review by jstrzebonski — APPROVED** [review]
Looks good

File: docker-compose.yml [hunk 3]
```diff
@@ -246,7 +247,9 @@ services:
       dockerfile: ci/docker/ubuntu-${UBUNTU}-${COMPILER_TYPE}-openmpi-cpp.dockerfile
       cache_from:
         - ${REPO}:${ARCH}-ubuntu-${UBUNTU}-${COMPILER}-openmpi-cpp
-      args: *default-args
+      args:
+        <<: *default-args
+        MPI_FLAGS: "--oversubscribe"
```
Comment (cz4rs) [inline docker-compose.yml:252]: can this be enabled unconditionally?

**Review by cz4rs — APPROVED** [review]
looks good

File: docker-compose.yml (hunk 3 above)
Comment (JacobDomagala) [reply docker-compose.yml]: I guess there's no harm in moving this flag to OpenMPI dockerfile

Comment (JacobDomagala) [reply docker-compose.yml]: Changed with https://github.com/DARMA-tasking/vt/pull/1306/commits/0a87130c583fda38c3d5d0eb03fa1a717bd2e94f

**Comment by JacobDomagala** [comment]
```
  224 - vt:*/TestLoadBalancer.test_load_balancer_1/*_proc_4 (Timeout)
  225 - vt:*/TestLoadBalancer.test_load_balancer_keep_last_elm/*_proc_4 (Timeout)
```
These started to timeout with latest changes to develop. I think I will disable the on 4 nodes

**Review by lifflander — APPROVED** [review]
Looks good

**Comment by JacobDomagala** [comment]
Ok so I've set 4 nodes for:
- Clang 10
- gcc 10 OpenMPI
- Intel 18 (extended)
- Intel 19
- NVCC 11 (both extended and standard)

Do we want to set it on any other workflow?

**Comment by nlslatt** [comment]
I submitted #1380 that should bring down the oversubscribed LB test duration enough to not time out.

**Comment by nlslatt** [comment]
> I submitted #1380 that should bring down the oversubscribed LB test duration enough to not time out.

I merged in #1380.  Please rebase and see if it helps you.

**Comment by JacobDomagala** [comment]
> > I submitted #1380 that should bring down the oversubscribed LB test duration enough to not time out.
> 
> I merged in #1380. Please rebase and see if it helps you.

Seems to be down to 30 sec now. I will keep it oversubscribed then.

**Pull request merged**